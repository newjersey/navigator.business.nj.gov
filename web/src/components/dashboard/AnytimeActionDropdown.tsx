import { HorizontalLine } from "@/components/HorizontalLine";
import { MenuOptionSelected } from "@/components/MenuOptionSelected";
import { MenuOptionUnselected } from "@/components/MenuOptionUnselected";
import { Heading } from "@/components/njwds-extended/Heading";
import { PrimaryButton } from "@/components/njwds-extended/PrimaryButton";
import { MediaQueries } from "@/lib/PageSizes";
import { useConfig } from "@/lib/data-hooks/useConfig";
import { useUserData } from "@/lib/data-hooks/useUserData";
import { ROUTES } from "@/lib/domain-logic/routes";
import {
  AnytimeActionLicenseReinstatement,
  AnytimeActionTask,
} from "@/lib/types/types";
import analytics from "@/lib/utils/analytics";
import { Autocomplete, TextField, useMediaQuery } from "@mui/material";
import { orderBy } from "lodash";
import { useRouter } from "next/compat/router";
import { ChangeEvent, type ReactElement, ReactNode, useState } from "react";

interface Props {
  anytimeActionTasks: AnytimeActionTask[];
  anytimeActionLicenseReinstatements: AnytimeActionLicenseReinstatement[];
}

type AnytimeAction = AnytimeActionTask | AnytimeActionLicenseReinstatement;
type AnytimeActionWithTypeAndCategory = AnytimeAction & {
  type: string;
  category: string[];
};

const getBoldedTextComponent = (
  searchValue: string,
  textToBold: string
): ReactNode => {
  const matches = textToBold.toLowerCase().indexOf(searchValue.toLowerCase());
  if (matches >= 0) {
    const beforeMatch = textToBold.slice(0, matches);
    const match = textToBold.slice(matches, matches + searchValue.length);
    const afterMatch = textToBold.slice(matches + searchValue.length);

    return (
      <>
        {beforeMatch}
        <span className="text-bold">{match}</span>
        {afterMatch}
      </>
    );
  }
  return <>{textToBold}</>;
};

export const AnytimeActionDropdown = (props: Props): ReactElement => {
  const { Config } = useConfig();
  const [selectedAnytimeAction, setSelectedAnytimeAction] = useState<
    AnytimeActionWithTypeAndCategory | undefined
  >(undefined);
  const router = useRouter();
  const { business } = useUserData();
  const isDesktopAndUp = useMediaQuery(MediaQueries.desktopAndUp);
  const industryId = business?.profileData.industryId;
  const sectorId = business?.profileData.sectorId;

  const getApplicableAnytimeActions =
    (): AnytimeActionWithTypeAndCategory[] => {
      let anytimeActionTasksWithType = props.anytimeActionTasks
        .filter((action) => findMatch(action))
        .map((action) => {
          return {
            ...action,
            type: "task",
            category: action.category,
          };
        });
      anytimeActionTasksWithType = orderBy(anytimeActionTasksWithType, [
        "name",
      ]);

      let anytimeActionLicenseReinstatementsWithType =
        props.anytimeActionLicenseReinstatements
          .filter((action) => licenseReinstatementMatch(action))
          .map((action) => {
            return {
              ...action,
              type: "license-reinstatement",
              category: [
                "Reactivate My Expired Permit, License or Registration",
              ],
            };
          });
      anytimeActionLicenseReinstatementsWithType = orderBy(
        anytimeActionLicenseReinstatementsWithType,
        ["name"]
      );

      let applicableAnytimeActions: AnytimeActionWithTypeAndCategory[] = [];
      applicableAnytimeActions.push(...anytimeActionTasksWithType);
      applicableAnytimeActions.push(
        ...anytimeActionLicenseReinstatementsWithType
      );
      applicableAnytimeActions = orderBy(applicableAnytimeActions, [
        "category",
      ]);

      return applicableAnytimeActions;
    };

  const findMatch = (action: AnytimeActionTask): boolean => {
    if ("category" in action && action.category[0] === "Only Show in Subtask")
      return false;
    if (action.applyToAllUsers) return true;
    if (
      action.industryIds &&
      industryId &&
      action.industryIds.includes(industryId)
    )
      return true;
    if (isAnytimeActionFromNonEssentialQuestions(action)) return true;
    return !!(
      action.sectorIds &&
      sectorId &&
      action.sectorIds.includes(sectorId)
    );
  };

  const isAnytimeActionFromNonEssentialQuestions = (
    action: AnytimeActionTask
  ): boolean => {
    switch (action.filename) {
      case "carnival-ride-supplemental-modification":
        return (
          !!business?.profileData.carnivalRideOwningBusiness ||
          !!business?.profileData.nonEssentialRadioAnswers[
            "carnival-ride-permitting"
          ]
        );
      case "operating-carnival-fire-permit":
        return (
          !!business?.profileData.carnivalRideOwningBusiness ||
          !!business?.profileData.travelingCircusOrCarnivalOwningBusiness ||
          !!business?.profileData.nonEssentialRadioAnswers[
            "carnival-ride-permitting"
          ] ||
          !!business?.profileData.nonEssentialRadioAnswers[
            "carnival-fire-licenses"
          ]
        );
      case "vacant-building-fire-permit":
        return !!business?.profileData.vacantPropertyOwner;
      default:
        return false;
    }
  };

  const licenseReinstatementMatch = (
    action: AnytimeActionLicenseReinstatement
  ): boolean => {
    const licenseNameFromAnytimeAction = action.licenseName;
    const licenseStatus =
      business?.licenseData?.licenses?.[licenseNameFromAnytimeAction]
        ?.licenseStatus;

    return licenseStatus === "EXPIRED";
  };

  const handleChange = (
    event: ChangeEvent<unknown>,
    value: AnytimeActionWithTypeAndCategory | null
  ): void => {
    if (value === null) {
      setSelectedAnytimeAction(undefined);
    } else {
      setSelectedAnytimeAction(value);
    }
  };

  return (
    <div className={"anytime-action-dropdown-container"}>
      <Heading level={2} className={"h2-styling text-medium"}>
        {Config.dashboardAnytimeActionDefaults.defaultHeaderText}
      </Heading>
      <HorizontalLine ariaHidden={true} />
      <div className="text-bold">
        {Config.dashboardAnytimeActionDefaults.defaultAutocompleteHeaderText}
      </div>
      <span className={isDesktopAndUp ? "flex" : "flex-column"}>
        <Autocomplete
          renderInput={(params): ReactElement => {
            return (
              <TextField
                {...params}
                inputProps={{
                  "aria-label": "anytimeActionDropdown",
                  "data-testid": "anytimeActionDropdown",
                  className: "",
                  ...params.inputProps,
                }}
                variant="outlined"
              />
            );
          }}
          componentsProps={{
            popper: {
              modifiers: [
                {
                  name: "flip",
                  enabled: false,
                },
              ],
            },
          }}
          getOptionLabel={(option: AnytimeActionWithTypeAndCategory) => {
            return option.name;
          }}
          isOptionEqualToValue={(option, value) => {
            return (
              option.name === value.name && option.filename === value.filename
            );
          }}
          groupBy={(option) => option.category[0]} // Currently just showing the first category
          renderGroup={(params) => (
            <li key={params.key} className="anytime-action-header-group">
              <div className="text-secondary-vivid text-bold padding-left-2 ">
                {params.group}
              </div>
              <ul className="anytime-action-dropdown-ul-list-container">
                {params.children}
              </ul>
            </li>
          )}
          options={getApplicableAnytimeActions()}
          filterOptions={(options, state) => {
            const searchValue = state.inputValue.toLowerCase();
            if (searchValue === "") {
              return options;
            }
            return options.filter((option) => {
              return (
                option.name.toLowerCase().includes(searchValue) ||
                option.description?.toLowerCase().includes(searchValue) ||
                option.searchMetaDataMatch?.toLowerCase().includes(searchValue)
              );
            });
          }}
          renderOption={(
            _props,
            option: AnytimeActionWithTypeAndCategory,
            { selected, inputValue }
          ): ReactElement => {
            const titleText = getBoldedTextComponent(inputValue, option.name);
            const descriptionText = getBoldedTextComponent(
              inputValue,
              option.description ?? ""
            );

            const newClassName = `${
              _props.className
            } anytime-action-dropdown-option ${
              selected ? "bg-accent-cool-lightest" : ""
            } fdc`;

            return (
              <li
                data-testid={`${option.filename}-option`}
                {..._props}
                className={newClassName}
                key={option.filename}
              >
                {selected ? (
                  <MenuOptionSelected secondaryText={descriptionText}>
                    {titleText}
                  </MenuOptionSelected>
                ) : (
                  <MenuOptionUnselected secondaryText={descriptionText}>
                    {titleText}
                  </MenuOptionUnselected>
                )}
              </li>
            );
          }}
          onChange={handleChange}
          className={
            selectedAnytimeAction
              ? `fg1 anytime-action-dropdown ${
                  !isDesktopAndUp && "margin-bottom-1"
                }`
              : " anytime-action-dropdown width-100"
          }
        />
        {selectedAnytimeAction && (
          <span
            className={
              isDesktopAndUp
                ? "anytime-action-primary-button"
                : "anytime-action-primary-button-mobile"
            }
          >
            {" "}
            <PrimaryButton
              isColor={"primary"}
              dataTestId={"anytimeActionPrimaryButton"}
              onClick={() => {
                if (!router) return;
                analytics.event.anytime_action_button.click.go_to_anytime_action_screen(
                  selectedAnytimeAction.filename
                );
                if (selectedAnytimeAction.type === "task") {
                  router.push(
                    `${ROUTES.anytimeActions}/${
                      (selectedAnytimeAction as AnytimeActionTask).urlSlug
                    }`
                  );
                }
                if (selectedAnytimeAction.type === "license-reinstatement") {
                  router.push(
                    `${ROUTES.licenseReinstatement}/${
                      (
                        selectedAnytimeAction as AnytimeActionLicenseReinstatement
                      ).urlSlug
                    }`
                  );
                }
              }}
            >
              {
                Config.dashboardAnytimeActionDefaults
                  .anytimeActionPageButtonText
              }
            </PrimaryButton>
          </span>
        )}
      </span>
    </div>
  );
};
