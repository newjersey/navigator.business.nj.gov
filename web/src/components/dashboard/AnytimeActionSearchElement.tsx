import { ReactElement, ReactNode } from "react";

interface Props {
  children: ReactNode;
  secondaryText?: ReactNode;
}

export const AnytimeActionSearchElement = (props: Props): ReactElement => {
  return (
    <>
      <div className={`padding-left-1 text-wrap padding-right-205`} data-testid="option">
        {props.children}
      </div>

      {props.secondaryText && (
        <div className={`padding-left-1 font-body-2xs text-base text-wrap`}>
          {props.secondaryText}
        </div>
      )}
    </>
  );
};
