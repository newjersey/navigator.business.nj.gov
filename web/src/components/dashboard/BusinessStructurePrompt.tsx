import { Content } from "@/components/Content";
import { PrimaryButton } from "@/components/njwds-extended/PrimaryButton";
import { useConfig } from "@/lib/data-hooks/useConfig";
import { ROUTES } from "@/lib/domain-logic/routes";
import { MediaQueries } from "@/lib/PageSizes";
import { useMediaQuery } from "@mui/material";
import { useRouter } from "next/router";
import { ReactElement } from "react";

interface Props {
  isButtonHidden?: boolean;
}
export const BusinessStructurePrompt = (props: Props): ReactElement => {
  const { Config } = useConfig();
  const isDesktopAndUp = useMediaQuery(MediaQueries.desktopAndUp);
  const router = useRouter();

  return (
    <div
      className={"business-structure-prompt-border padding-3 margin-top-2"}
      data-testid={"business-structure-prompt"}
    >
      <div className={isDesktopAndUp ? "flex flex-row flex-align-start" : ""}>
        <div className={isDesktopAndUp ? "flex-fill margin-right-2" : "margin-bottom-2"}>
          {router.asPath === ROUTES.businessStructureTask ? (
            <div data-testid={"business-structure-task-content"}>
              <Content>{Config.businessStructurePrompt.notCompletedTaskPromptBusinessStructureTask}</Content>
            </div>
          ) : (
            <div data-testid={"any-task-content"}>
              <Content>{Config.businessStructurePrompt.notCompletedTaskPromptAnyTask}</Content>
            </div>
          )}
        </div>
        {!props.isButtonHidden && (
          <div className={isDesktopAndUp ? "flex-auto" : ""}>
            <PrimaryButton
              isColor="primary"
              isRightMarginRemoved
              isFullWidthOnDesktop
              onClick={(): void => {
                router.push(ROUTES.businessStructureTask);
              }}
              dataTestId={"business-structure-prompt-button"}
            >
              {Config.businessStructurePrompt.buttonText}
            </PrimaryButton>
          </div>
        )}
      </div>
    </div>
  );
};
