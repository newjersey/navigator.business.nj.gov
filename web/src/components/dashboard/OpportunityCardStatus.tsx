import { toProperCase } from "@businessnjgovnavigator/shared/";
import React from "react";

interface Props {
  dueDate?: string;
  status?: string;
}

export const OpportunityCardStatus = (props: Props) => {
  return (
    <>
      {props.dueDate && props.status !== "first come, first serve" && props.status !== "rolling application" && (
        <div className="dashboard-opportunity-card-due-date">
          <span className="dashboard-opportunity-card-due-date-header">Due: </span>
          {props.dueDate}{" "}
        </div>
      )}
      {(props.status === "first come, first serve" || props.status === "rolling application") && (
        <div className="dashboard-opportunity-card-due-date">{toProperCase(props.status)}</div>
      )}
    </>
  );
};
