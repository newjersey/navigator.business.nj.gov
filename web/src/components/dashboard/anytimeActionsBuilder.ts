import { getNonEssentialQuestionAnytimeActions } from "@/lib/domain-logic/getNonEssentialQuestionAnytimeActions";
import { ProfileData } from "@businessnjgovnavigator/shared/profileData";
import { AnytimeActionTask } from "@businessnjgovnavigator/shared/types";

export const getAnytimeActionsFromNonEssentialQuestions = (
  profileData: ProfileData | undefined,
  allAnytimeActionTasks: AnytimeActionTask[],
): AnytimeActionTask[] => {
  if (!profileData) return [];
  if (Object.keys(profileData.nonEssentialRadioAnswers).length === 0) return [];

  const anytimeActionIds = anytimeActionsToAdd(profileData.nonEssentialRadioAnswers);
  const anytimeActionsFromNonEssentialQuestions = allAnytimeActionTasks.filter((anytimeAction) =>
    anytimeActionIds.includes(anytimeAction.filename),
  );

  return anytimeActionsFromNonEssentialQuestions;
};

const anytimeActionsToAdd = (
  nonEssentialRadioAnswers: Record<string, boolean | undefined>,
): string[] => {
  let anytimeActionIds: string[] = [];

  for (const [questionId, response] of Object.entries(nonEssentialRadioAnswers)) {
    if (response) {
      const anytimeActionId = getNonEssentialQuestionAnytimeActions(questionId);
      anytimeActionIds = [...anytimeActionIds, ...anytimeActionId];
    }
  }
  return anytimeActionIds;
};
