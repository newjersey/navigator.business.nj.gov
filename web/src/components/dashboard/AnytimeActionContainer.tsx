import { AnytimeActionSearch } from "@/components/dashboard/AnytimeActionSearch";
import { HorizontalLine } from "@/components/HorizontalLine";
import { Heading } from "@/components/njwds-extended/Heading";
import { useConfig } from "@/lib/data-hooks/useConfig";
import {
  AnytimeActionLicenseReinstatement,
  AnytimeActionTask,
} from "@businessnjgovnavigator/shared/types";
import { ReactElement } from "react";

interface Props {
  commonBusinessTasks: (AnytimeActionLicenseReinstatement | AnytimeActionTask)[];
  anytimeActionTasks: AnytimeActionTask[];
  anytimeActionTasksFromNonEssentialQuestions: AnytimeActionTask[];
  anytimeActionLicenseReinstatements: AnytimeActionLicenseReinstatement[];
}

export const AnytimeActionContainer = (props: Props): ReactElement => {
  const { Config } = useConfig();

  return (
    <div className={"anytime-action-dropdown-container"}>
      <Heading level={2}>{Config.dashboardAnytimeActionDefaults.defaultHeaderText}</Heading>
      <HorizontalLine ariaHidden={true} />
      <Heading level={3}>{Config.dashboardAnytimeActionDefaults.searchFieldHeaderText}</Heading>
      <AnytimeActionSearch
        anytimeActionTasks={props.anytimeActionTasks}
        anytimeActionTasksFromNonEssentialQuestions={
          props.anytimeActionTasksFromNonEssentialQuestions
        }
        anytimeActionLicenseReinstatements={props.anytimeActionLicenseReinstatements}
      />
      {props.commonBusinessTasks.length > 0 && (
        <>
          <Heading level={3}>
            {Config.dashboardAnytimeActionDefaults.commonBusinessTasksHeader}
          </Heading>
          <ul>
            {props.commonBusinessTasks.map((businessTask) => (
              <li key={businessTask.urlSlug}>{businessTask.name}</li>
            ))}
          </ul>
        </>
      )}
    </div>
  );
};
