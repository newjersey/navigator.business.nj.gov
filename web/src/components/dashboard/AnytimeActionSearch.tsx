import { Content } from "@/components/Content";
import { AnytimeActionSearchElement } from "@/components/dashboard/AnytimeActionSearchElement";
import { MediaQueries } from "@/lib/PageSizes";
import { useConfig } from "@/lib/data-hooks/useConfig";
import { useUserData } from "@/lib/data-hooks/useUserData";
import { ROUTES } from "@/lib/domain-logic/routes";
import analytics from "@/lib/utils/analytics";
import { templateEval } from "@/lib/utils/helpers";
import {
  AnytimeActionCategory,
  AnytimeActionLicenseReinstatement,
  AnytimeActionTask,
} from "@businessnjgovnavigator/shared/types";
import { Autocomplete, TextField, useMediaQuery } from "@mui/material";
import { orderBy, unionBy } from "lodash";
import { useRouter } from "next/compat/router";
import { type ReactElement, ReactNode, useState } from "react";

interface Props {
  anytimeActionTasks: AnytimeActionTask[];
  anytimeActionTasksFromNonEssentialQuestions: AnytimeActionTask[];
  anytimeActionLicenseReinstatements: AnytimeActionLicenseReinstatement[];
}

type AnytimeAction = AnytimeActionTask | AnytimeActionLicenseReinstatement;
type AnytimeActionWithTypeAndCategory = AnytimeAction & {
  type: string;
  category: AnytimeActionCategory[];
};

const getBoldedTextComponent = (searchValue: string, textToBold: string): ReactNode => {
  const matches = textToBold.toLowerCase().indexOf(searchValue.toLowerCase());
  if (matches >= 0) {
    const beforeMatch = textToBold.slice(0, matches);
    const match = textToBold.slice(matches, matches + searchValue.length);
    const afterMatch = textToBold.slice(matches + searchValue.length);

    return (
      <>
        {beforeMatch}
        <span className="text-bold">{match}</span>
        {afterMatch}
      </>
    );
  }
  return <>{textToBold}</>;
};

export const AnytimeActionSearch = (props: Props): ReactElement => {
  const { Config } = useConfig();
  const router = useRouter();
  const { business } = useUserData();
  const isDesktopAndUp = useMediaQuery(MediaQueries.desktopAndUp);
  const industryId = business?.profileData.industryId;
  const sectorId = business?.profileData.sectorId;

  const getApplicableAnytimeActions = (): AnytimeActionWithTypeAndCategory[] => {
    const anytimeActionTasks = moveAnytimeActionsToRecommendedForYouSection(
      props.anytimeActionTasks,
    );

    const [tasks, reinstatements] = [
      anytimeActionTasks.filter(findMatch).map((action) => {
        return {
          ...action,
          category: action.category,
        };
      }),
      props.anytimeActionLicenseReinstatements.filter(licenseReinstatementMatch).map((action) => {
        return {
          ...action,
        };
      }),
    ];

    const deduplicatedTasks = unionBy(
      tasks,
      props.anytimeActionTasksFromNonEssentialQuestions,
      "name",
    );

    const orderedTasks = orderBy(deduplicatedTasks, ["name"]);
    const orderedReinstatements = orderBy(reinstatements, ["name"]);

    const allActions = [...orderedTasks, ...orderedReinstatements];

    return allActions.sort((a, b) => {
      const categoryA = a.category[0].categoryName.toLowerCase();
      const categoryB = b.category[0].categoryName.toLowerCase();

      if (categoryA === RECOMMENDED_FOR_YOU_DISPLAY_TEXT.toLowerCase()) {
        return -1;
      }
      if (categoryB === RECOMMENDED_FOR_YOU_DISPLAY_TEXT.toLowerCase()) {
        return 1;
      }

      return categoryA.localeCompare(categoryB);
    });
  };

  const findMatch = (action: AnytimeActionTask): boolean => {
    if ("category" in action && action.category[0].categoryId === "only-show-in-subtask")
      return false;
    if (action.applyToAllUsers) return true;
    if (action.industryIds && industryId && action.industryIds.includes(industryId)) return true;
    if (isAnytimeActionFromNonEssentialQuestions(action)) return true;
    return !!(action.sectorIds && sectorId && action.sectorIds.includes(sectorId));
  };

  const isAnytimeActionFromNonEssentialQuestions = (action: AnytimeActionTask): boolean => {
    switch (action.filename) {
      case "carnival-ride-supplemental-modification":
        return (
          !!business?.profileData.carnivalRideOwningBusiness ||
          !!business?.profileData.nonEssentialRadioAnswers["carnival-ride-permitting"]
        );
      case "operating-carnival-fire-permit":
        return (
          !!business?.profileData.carnivalRideOwningBusiness ||
          !!business?.profileData.travelingCircusOrCarnivalOwningBusiness ||
          !!business?.profileData.nonEssentialRadioAnswers["carnival-ride-permitting"] ||
          !!business?.profileData.nonEssentialRadioAnswers["carnival-fire-licenses"]
        );
      case "vacant-building-fire-permit":
        return !!business?.profileData.vacantPropertyOwner;
      default:
        return false;
    }
  };

  const licenseReinstatementMatch = (action: AnytimeActionLicenseReinstatement): boolean => {
    const licenseNameFromAnytimeAction = action.licenseName;
    const licenseStatus =
      business?.licenseData?.licenses?.[licenseNameFromAnytimeAction]?.licenseStatus;

    return licenseStatus === "EXPIRED";
  };

  const RECOMMENDED_FOR_YOU_DISPLAY_TEXT =
    Config.dashboardAnytimeActionDefaults.recommendedForYouCategoryHeader;
  const RECOMMENDED_FOR_YOU_ID = "recommended-for-you";

  const moveAnytimeActionsToRecommendedForYouSection = (
    anytimeActions: AnytimeActionTask[],
  ): AnytimeActionTask[] => {
    const anytimeActionsWithRecommendedForYouCategorysOverriden = anytimeActions;

    for (const [
      index,
      anytimeAction,
    ] of anytimeActionsWithRecommendedForYouCategorysOverriden.entries()) {
      if (anytimeAction.moveToRecommendedForYouSection) {
        anytimeActionsWithRecommendedForYouCategorysOverriden[index].category[0].categoryName =
          RECOMMENDED_FOR_YOU_DISPLAY_TEXT;
        anytimeActionsWithRecommendedForYouCategorysOverriden[index].category[0].categoryId =
          RECOMMENDED_FOR_YOU_ID;
      }
      if (anytimeAction.nonEssentialQuestionsMoveToRecommendedAnytimeActionIds) {
        for (const nonEssentialQuestionId of anytimeAction.nonEssentialQuestionsMoveToRecommendedAnytimeActionIds) {
          if (business?.profileData.nonEssentialRadioAnswers[nonEssentialQuestionId]) {
            anytimeActionsWithRecommendedForYouCategorysOverriden[index].category[0].categoryName =
              RECOMMENDED_FOR_YOU_DISPLAY_TEXT;
            anytimeActionsWithRecommendedForYouCategorysOverriden[index].category[0].categoryId =
              RECOMMENDED_FOR_YOU_ID;
          }
        }
      }
    }

    return anytimeActionsWithRecommendedForYouCategorysOverriden;
  };

  const [isFocused, setIsFocused] = useState(false);
  const [open, setOpen] = useState(false);
  const [inputValue, setInputValue] = useState("");

  const handleAnytimeActionClick = (value: AnytimeActionWithTypeAndCategory): void => {
    if (router && value && typeof value !== "string") {
      analytics.event.anytime_action_button.click.go_to_anytime_action_screen(value.filename);
      if (value.type === "task") {
        router.push(`${ROUTES.anytimeActions}/${(value as AnytimeActionTask).urlSlug}`);
      }
      if (value.type === "license-reinstatement") {
        router.push(
          `${ROUTES.licenseReinstatement}/${(value as AnytimeActionLicenseReinstatement).urlSlug}`,
        );
      }
    }
  };

  return (
    <>
      <span className={isDesktopAndUp ? "flex" : "flex-column"}>
        <Autocomplete
          open={open}
          onOpen={() => setOpen(true)}
          onClose={() => setOpen(false)}
          renderInput={(params): ReactElement => {
            return (
              <TextField
                {...params}
                inputProps={{
                  "aria-label": "anytimeActionSearch",
                  "data-testid": "anytimeActionSearch",
                  ...params.inputProps,
                  className: `disable-focus-outline ${params.inputProps.className}`,
                }}
                placeholder={
                  isFocused ? "" : Config.dashboardAnytimeActionDefaults.searchFieldHintText
                }
                onFocus={() => {
                  setIsFocused(true);
                  setOpen(true);
                }}
                onBlur={() => setIsFocused(false)}
                InputProps={{
                  ...params.InputProps,
                  startAdornment: (
                    <img className="" src="/img/search.svg" alt="" role="presentation" />
                  ),
                }}
                variant="outlined"
              />
            );
          }}
          onInputChange={(event, newInputValue) => {
            setInputValue(newInputValue);
          }}
          inputValue={inputValue}
          componentsProps={{
            popper: {
              modifiers: [
                {
                  name: "flip",
                  enabled: false,
                },
              ],
            },
          }}
          getOptionLabel={(option: AnytimeActionWithTypeAndCategory | string) => {
            if (typeof option === "string") {
              return option;
            }
            return option.name;
          }}
          isOptionEqualToValue={(option, value) => {
            return option.name === value.name && option.filename === value.filename;
          }}
          groupBy={(option) => option.category[0].categoryName} // Currently just showing the first category
          renderGroup={(params) => {
            return (
              <li key={params.key} className="anytime-action-header-group">
                <div className="padding-left-2 font-body-2xs text-base">{params.group}</div>
                <ul className="anytime-action-dropdown-ul-list-container">{params.children}</ul>
              </li>
            );
          }}
          options={getApplicableAnytimeActions()}
          filterOptions={(options, state) => {
            const searchValue = state.inputValue.toLowerCase();
            if (searchValue === "") {
              return options;
            }
            return options.filter((option) => {
              if (option.synonyms) {
                for (const synonym of option.synonyms) {
                  if (synonym.toLowerCase().includes(searchValue)) {
                    return true;
                  }
                }
              }
              return (
                option.name.toLowerCase().includes(searchValue) ||
                option.description?.toLowerCase().includes(searchValue)
              );
            });
          }}
          onChange={(event, value) => {
            if (value !== null) {
              handleAnytimeActionClick(value);
            }
          }}
          noOptionsText={
            <Content>
              {templateEval(Config.dashboardAnytimeActionDefaults.searchHasNoOptionsText, {
                inputValue: inputValue,
              })}
            </Content>
          }
          renderOption={(
            _props,
            option: AnytimeActionWithTypeAndCategory,
            { inputValue },
          ): ReactElement => {
            const titleText = getBoldedTextComponent(inputValue, option.name);
            const descriptionText = getBoldedTextComponent(inputValue, option.description ?? "");

            return (
              // must fix for keyboard
              // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events
              <li
                data-testid={`${option.filename}-option`}
                {..._props}
                className={`${_props.className} anytime-action-dropdown-option fdr`}
                key={option.filename}
                onClick={() => {
                  handleAnytimeActionClick(option);
                }}
                onKeyDown={(event: React.KeyboardEvent) => {
                  if (event.key === "Enter") {
                    handleAnytimeActionClick(option);
                  }
                }}
              >
                <img
                  src="/img/subdirectory_arrow_right.svg"
                  alt=""
                  role="presentation"
                  key={"subdirectory_arrow"}
                />
                <div className="fdc" key={option.filename}>
                  <AnytimeActionSearchElement secondaryText={descriptionText}>
                    {titleText}
                  </AnytimeActionSearchElement>
                </div>
              </li>
            );
          }}
          className={" anytime-action-dropdown width-100 "}
        />
      </span>
    </>
  );
};
