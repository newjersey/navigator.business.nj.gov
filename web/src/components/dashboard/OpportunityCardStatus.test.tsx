import { OpportunityCardStatus } from "@/components/dashboard/OpportunityCardStatus";
import { render } from "@testing-library/react";

describe("OpportunityCard", () => {
  it("matches the layout", () => {
    const view = render(<OpportunityCardStatus />).baseElement;
    expect(view).toMatchSnapshot();
  });
  it("matches the layout with a due date", () => {
    const view = render(<OpportunityCardStatus dueDate="09/01/2030" />).baseElement;
    expect(view).toMatchSnapshot();
  });
  it("matches the layout when status is first come first serve", () => {
    const view = render(<OpportunityCardStatus status="first come, first serve" />).baseElement;
    expect(view).toMatchSnapshot();
  });
  it("matches the layout when status is rolling application", () => {
    const view = render(<OpportunityCardStatus status="rolling application" />).baseElement;
    expect(view).toMatchSnapshot();
  });
  it("renders with due date correctly", () => {
    const view = render(<OpportunityCardStatus dueDate="09/01/2030" />).baseElement;
    expect(view).toHaveTextContent("Due:");
    expect(view).toHaveTextContent("09/01/2030");
  });
  it("renders with text of First Come First Serve correctly", () => {
    const view = render(<OpportunityCardStatus status="first come, first serve" />).baseElement;
    expect(view).toContainHTML("First Come, First Serve");
  });
  it("does not render due date if status is first come, first serve", () => {
    const view = render(
      <OpportunityCardStatus dueDate="09/03/30" status="first come, first serve" />
    ).baseElement;
    expect(view).not.toHaveTextContent("Due:");
    expect(view).toContainHTML("First Come, First Serve");
  });
  it("renders with text of Rolling Application correctly", () => {
    const view = render(<OpportunityCardStatus status="rolling application" />).baseElement;
    expect(view).toContainHTML("Rolling Application");
  });
  it("does not render due date if status is rolling application", () => {
    const view = render(
      <OpportunityCardStatus dueDate="09/03/30" status="rolling application" />
    ).baseElement;
    expect(view).not.toHaveTextContent("Due:");
    expect(view).toContainHTML("Rolling Application");
  });
});
