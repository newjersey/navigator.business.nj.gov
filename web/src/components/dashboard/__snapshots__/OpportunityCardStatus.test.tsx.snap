// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpportunityCard matches the layout 1`] = `
<body>
  <div />
</body>
`;

exports[`OpportunityCard matches the layout when status is first come first serve 1`] = `
<body>
  <div>
    <div
      class="dashboard-opportunity-card-due-date text-bold"
    >
      First Come, First Serve
    </div>
  </div>
</body>
`;

exports[`OpportunityCard matches the layout when status is rolling application 1`] = `
<body>
  <div>
    <div
      class="dashboard-opportunity-card-due-date text-bold"
    >
      Rolling Application
    </div>
  </div>
</body>
`;

exports[`OpportunityCard matches the layout with a due date 1`] = `
<body>
  <div>
    <div
      class="dashboard-opportunity-card-due-date"
    >
      <span
        class="dashboard-opportunity-card-due-date-header"
      >
        Due: 
      </span>
      09/01/2030
       
    </div>
  </div>
</body>
`;
